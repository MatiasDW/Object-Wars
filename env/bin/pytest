#!/home/joan/Documents/git_repositories/juegoPOO/env/bin/python3

# -*- coding: utf-8 -*-
import re
import sys

from pytest import console_main

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(console_main())
