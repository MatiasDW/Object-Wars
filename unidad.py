from abc import ABC, abstractmethod

MAX_VIDA = 10


class Unidad():
    """Clase abstracta que modela una unidad"""

    def __init__(self):
        """Creadora del objecto Unidad"""
        pass

    def descansar(self):
        """ Metodo abastracto, restaura puntos de vida a la unidad"""
        pass

    def atacar(self):
        """Este metodo debe ser usado para consultar los puntos de ataque, en caso de que la unidad este atacando"""
        pass


class Soldado():
    """Unidad soldado, tiene un coste de 5 monedas, tiene 3 puntos de ataque y restaura 5 puntos de vida al descansar"""
     def __init__(self):
        super().__init__()
        self.coste = 5
        self.puntos_vida = MAX_VIDA
        self.puntos_ataque = 3

    def descansar(self):
        self.puntos_vida += 5


class Arquero():
    """ Unidad Arquero, tiene un coste de 6 monedas, tiene 8 puntos de ataque y restaura 2 puntos de vida al decansar
    Los arqueros atacan 1 de cada 2 veces ya que deben recargar, empiezan la partida sin estar preparados para atacar"""
    def __init__(self):
            super().__init__()
            self.coste = 6
            self.puntos_vida = MAX_VIDA
            self.puntos_ataque = 8
            self.preparado = False

        def descansar(self):
            self.puntos_vida += 2
            self.preparado = True

        def atacar(self):
            if self.preparado:
                self.preparado = False
                return self.puntos_ataque
            else:
                return 0

class Caballero():
    """ Unidad Caballero, tiene un coste de 9 monedas, tiene 5 puntos de ataque, y al descansar no restaura puntos de vida"""
     def __init__(self):
        super().__init__()
        self.coste = 9
        self.puntos_vida = MAX_VIDA
        self.puntos_ataque = 5

    def descansar(self):
        # Los caballeros no restauran puntos de vida al descansar
        pass

    def atacar(self):
        return self.puntos_ataque