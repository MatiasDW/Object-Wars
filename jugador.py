
from operator import concat
from unidad import *


class Jugador():
    "Un Jugador tiene un nombre, puntos_vida, monedas y unidades. Un jugador no puede deudas, es decir, no puede tener un numero de monedas negativo"

    def __init__(self, nombre, puntos_vida=20):
        pass

    def descansar(self):
        """Hace que la primera unidad, si la hay, descanse"""
        pass

    def get_monedas(self):
        """Devuelve el numero de monedas actual del jugador"""
        pass

    def set_monedas(self, value):
        """Modifica el numero de monedas por el valor value"""
        pass
