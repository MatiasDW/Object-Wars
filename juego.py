import os
from jugador import Jugador
from unidad import Soldado, Arquero, Caballero
import sys

N_VIDAS = 20
MONEDAS_TURNO = 10
BONUS_DANO = 1.5


class Juego():

    def __init__(self, nombre_jugador1,  nombre_jugador2):
        """Creadora"""
        self.jugador1 = Jugador(nombre=nombre_jugador1,
                                puntos_vida=20)
        self.jugador2 = Jugador(nombre=nombre_jugador2,
                                puntos_vida=20)

    @staticmethod
    def _elegir_opcion(options):
        """Muestra por pantalla la lista de opciones enumeradas y retorna el número de opción elegida.
         options es una lista de strings"""
        print("Elige una opción:")
        for idx, element in enumerate(options):
            print("{}) {}".format(idx+1, element))
        i = input("Introduce un número: ")
        try:
            if 0 < int(i) <= len(options):
                return int(i)
        except:
            pass
        return None

    def _turno(self, jugador):
        f"""Se le añaden {MONEDAS_TURNO} al jugador, se le muestran las opciones de compra hasta que decida finalizar el turno"""
        pass

    def _finalizar(self, jugador_ganador):
        """Finaliza la partida mostrando como ganador al jugador_ganador"""
        self._clear_screen()

        print(f"""FIN DEL JUEGO
El jugador ganador es {jugador_ganador.nombre}""")
        quit()

    @staticmethod
    def _clear_screen():
        if(os.name == 'posix'):
            os.system('clear')
        # else screen will be cleared for windows
        else:
            os.system('cls')
            pass

    def loop(self):
        "Loop del juego, se ejecuta hasta finalizar la partida"
        ronda = 0
        while(True):
            ronda += 1
            self._turno(self.jugador1)
            self._clear_screen()
            self._turno(self.jugador2)
            self._batalla()

            # limitación para permitir tests de funcionalidad sin implementación
            if ronda == 100:
                break

    def _daño_al_jugador(self, defensor, atacante):
        """
        El defensor recibe un ataque de cada unidad del atacante, si el defensor se queda sin
        puntos de vida, llama a la función _finalizar.
        """
        pass

    @staticmethod
    def _calcular_bonus(atacante, defensor):
        """Devuelve el bonus de ataque que tiene el atacante contra el defensor, se debe usar
         la función isinstace(instancia, clase) para implementarla"""

    def _batalla(self):
        """Realiza una batalla entre las unidades del jugador1 y el jugador2. 
        La batalla se desarrolla en combates 1 vs 1, siempre entre las unidades más antiguas de cada jugador.
        Durante un combate, las dos unidades pierdan tantos puntos de vida como puntos de ataque tenga la unidad adversaria.
        Si una unidad sobrevive a un combate, ésta participará en el siguiente combate, la batalla continua mientras ambos jugadores
         tengan unidades disponibles. 
        Cuando a un jugador no le queden más unidades, recibe un ataque de cada unidad enemiga remaniente.
        Al acabar, los jugadores hacen descansar a sus unidades.
        """
        pass

    @classmethod
    def mensaje_bienvenida(self):
        print(f"""
Bienvenido a ObjectWars
Este juego es un juego para dos jugadores. Lo que un jugador realiza durante su turno es secreto, por lo que el otro jugador no debe mirar las acciones que realize el otro \
jugador durante su turno.

El objectivo del juego es dejar el enemigo sin puntos de vida, al empezar cada jugador dispone de {N_VIDAS}.

Durante un turno el jugador puede comprar unidades.

Al empejar un turno cada jugador recibe {MONEDAS_TURNO} mondedas.

Despues de que ambos jugadores acaben sus turnos sus unidades entraran en combate. Si un jugador no tiene unidades, o son derrotadas, recibirá el daño en sus puntos de vida.

Existen tres tipologias de unidades: soldados, arqueros y caballeros. Las unidades tienen un bonus de {BONUS_DANO} de daño siguiendo la siguiente jerarquía:
soldado -> arquero -> caballero -> soldado


    
    """)


if __name__ == "__main__":
    Juego.mensaje_bienvenida()
    nombre1 = input("Introduce el nombre del jugador1: ")
    nombre2 = input("Introduce el nombre del jugador2: ")
    juego = Juego(nombre1, nombre2)
    juego.loop()
