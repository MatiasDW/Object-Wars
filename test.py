import sys
import pytest
from juego import Juego, N_VIDAS
from unidad import MAX_VIDA, Arquero, Caballero, Soldado, Unidad
from jugador import Jugador
from os import path
import shutil

import filecmp

"""
Cuando esta constante es true, substituirá los outputs de los tests de funcionalidad
por los actuales y ara fallar los tests hasta que se cambie de nuevo
"""
CREAR_OUTPUTS_FUNCIONALIDAD = False


class Test_unitarios():
    def test_crear_jugador(self):
        mi_jugador = Jugador("Bobby Fischer")
        mi_jugador.get_monedas()
        assert(mi_jugador.nombre == "Bobby Fischer")
        assert(mi_jugador.unidades == [])
        assert(mi_jugador.puntos_vida > 0)

    def test_monedas_negativas_jugador(self):
        mi_jugador = Jugador("José Raúl Capablanca")
        mi_jugador.set_monedas(-1)
        assert mi_jugador.get_monedas() == 0

    def test_crear_caballero(self):
        mi_caballero = Caballero()
        with pytest.raises(AttributeError):
            mi_caballero.preparado

    def test_herencia(self):
        assert(issubclass(Soldado, Unidad))
        assert(issubclass(Arquero, Unidad))
        assert(issubclass(Caballero, Unidad))

    def test_abstract_class(self):
        with pytest.raises(TypeError):
            mi_unidad = Unidad()

    def test_ataque_arquero(self):
        mi_arquero = Arquero()
        assert(mi_arquero.atacar() == 0)
        assert(mi_arquero.atacar() > 0)
        assert(mi_arquero.atacar() == 0)
        assert(mi_arquero.atacar() > 0)

    def test_ataque_soldado(self):
        mi_soldado = Soldado()
        assert(mi_soldado.atacar() > 0)

    def test_descansar_soldado(self):
        mi_soldado = Soldado()
        max_puntos_vida = mi_soldado.puntos_vida
        mi_soldado.puntos_vida -= 1
        mi_soldado.descansar()

        assert(max_puntos_vida == mi_soldado.puntos_vida)

    def test_bonus(self):
        miJuego = Juego("Jugador1", "Jugador2")
        miSoldado = Soldado()
        miArquero = Arquero()
        miCaballero = Caballero()

        assert(miJuego._calcular_bonus(miSoldado, miArquero) == 1.5)
        assert(miJuego._calcular_bonus(miArquero, miCaballero) == 1.5)
        assert(miJuego._calcular_bonus(miCaballero, miSoldado) == 1.5)
        assert(miJuego._calcular_bonus(miArquero, miSoldado) == 1)
        assert(miJuego._calcular_bonus(miCaballero, miCaballero) == 1)


class Test_integracion():

    def test_arquero_daño_jugador(self):
        miJuego = Juego("Jugador1", "Jugador2")

        miJuego.jugador2.unidades.append(Arquero())

        miJuego._batalla()
        miJuego._batalla()

        assert(miJuego.jugador1.puntos_vida == 12)
        assert(miJuego.jugador2.puntos_vida == 20)

    def test_descansar_jugador(self):
        mi_jugador = Jugador("Garry Kasparov")
        mi_soldado = Soldado()
        max_puntos_vida = mi_soldado.puntos_vida
        mi_soldado.puntos_vida -= 1
        mi_jugador.unidades.append(mi_soldado)
        mi_jugador.descansar()

        assert(max_puntos_vida == mi_soldado.puntos_vida)

    def test_descansar_jugador_sin_unidades(self):
        mi_jugador = Jugador("Garry Kasparov")
        mi_soldado = Soldado()
        mi_jugador.descansar()

    def test_daño_al_jugador(self):
        miJuego = Juego("Jugador1", "Jugador2")
        miSoldado = Soldado()
        miJuego.jugador1.unidades.append(miSoldado)
        miJuego._daño_al_jugador(miJuego.jugador2, miJuego.jugador1)
        assert(N_VIDAS-miSoldado._puntos_ataque ==
               miJuego.jugador2.puntos_vida)

    def test_finalizar_partida_jugador(self):
        miJuego = Juego("Jugador1", "Jugador2")
        for i in range(100):
            miJuego.jugador1.unidades.append(Soldado())
        with pytest.raises(SystemExit) as pytest_wrapped_e:
            miJuego._daño_al_jugador(
                miJuego.jugador2, miJuego.jugador1)

    def test_batalla_soldado_soldado(self):
        miJuego = Juego("Jugador1", "Jugador2")

        miSoldado = Soldado()
        miJuego.jugador1.unidades.append(miSoldado)

        miArquero = Soldado()
        miJuego.jugador2.unidades.append(miArquero)

        miJuego._batalla()
        assert(miJuego.jugador1.puntos_vida == 20)
        assert(miJuego.jugador2.puntos_vida == 20)

    def test_batalla_soldado_vs_arquero(self):
        miJuego = Juego("Jugador1", "Jugador2")

        miSoldado = Soldado()
        miJuego.jugador1.unidades.append(miSoldado)

        miArquero = Arquero()
        miJuego.jugador2.unidades.append(miArquero)

        miJuego._batalla()
        assert(miJuego.jugador1.puntos_vida == 20)
        assert(miJuego.jugador2.puntos_vida == 17)

    def test_batalla_soldado_vs_2_soldados(self):
        miJuego = Juego("Jugador1", "Jugador2")

        miJuego.jugador1.unidades.append(Soldado())

        miJuego.jugador2.unidades.append(Soldado())
        miJuego.jugador2.unidades.append(Soldado())

        miJuego._batalla()
        assert(miJuego.jugador1.puntos_vida == 17)
        assert(miJuego.jugador2.puntos_vida == 20)

    def test_batalla_soldado_vs_2_arqueros(self):
        miJuego = Juego("Jugador1", "Jugador2")

        miSoldado = Soldado()
        miJuego.jugador1.unidades.append(miSoldado)

        miJuego.jugador2.unidades.append(Arquero())
        miJuego.jugador2.unidades.append(Arquero())

        miJuego._batalla()
        assert(miJuego.jugador1.puntos_vida == 20)
        assert(miJuego.jugador2.puntos_vida == 20)


def check_same_output(path_orginial, path_test, create_output):
    """Devuelve true en caso de que los dos ficheros sean iguales, en caso de que create_output
    sea cierto copiar el output actual como el nuevo output a comparar, y hará fallar el test"""
    if create_output:
        shutil.copy(path_test, path_orginial)
        pytest.fail(f"Fichero {path_orginial} creado")
    
    
    with open(path_orginial) as file_1:
        file_1_text = file_1.readlines()
 
    with open(path_test) as file_2:
        file_2_text = file_2.readlines()
    
    assert len(file_1_text) == len(file_2_text)
    for line1, line2 in zip(file_1_text, file_2_text):
        assert line1 == line2



class Test_funcionalidad():

    def test_comprar_soldado(self, monkeypatch):
        sol_file_path = 'test/test_comprar_soldado.out'
        test_path = 'test/test.out'
        with open('test/test.out', 'w') as file:
            sys.stdout = file
            Juego.mensaje_bienvenida()
            juego = Juego("Jugador1", "jugador2")
            # Entrada ficticia de los jugadores
            answers = iter([3, 1])

            # using lambda statement for mocking
            monkeypatch.setattr('builtins.input', lambda name: next(answers))
            with pytest.raises(SystemExit) as e:
                juego.loop()
            assert e.type == SystemExit

        check_same_output(sol_file_path, test_path,
                          CREAR_OUTPUTS_FUNCIONALIDAD)

    def test_dinero_insuficiente(self, monkeypatch):
        sol_file_path = 'test/test_dinero_insuficiente.out'
        test_path = 'test/test.out'
        with open('test/test.out', 'w') as file:
            sys.stdout = file
            Juego.mensaje_bienvenida()
            juego = Juego("Jugador1", "jugador2")
            # Entrada ficticia de los jugadores
            answers = iter([3, 3, 3, 1])

            # using lambda statement for mocking
            monkeypatch.setattr('builtins.input', lambda name: next(answers))
            with pytest.raises(SystemExit) as e:
                juego.loop()
            assert e.type == SystemExit

        check_same_output(sol_file_path, test_path,
                          CREAR_OUTPUTS_FUNCIONALIDAD)

    def test_comprar_dos_soldados(self, monkeypatch):
        sol_file_path = 'test/test_comprar_dos_soldados.out'
        test_path = 'test/test.out'
        with open(test_path, 'w') as file:
            sys.stdout = file
            Juego.mensaje_bienvenida()
            juego = Juego("Jugador1", "jugador2")
            # Entrada ficticia de los jugadores
            answers = iter([3, 3, 1])

            # using lambda statement for mocking
            monkeypatch.setattr('builtins.input', lambda name: next(answers))
            with pytest.raises(SystemExit) as e:
                juego.loop()
            assert e.type == SystemExit

        check_same_output(sol_file_path, test_path,
                          CREAR_OUTPUTS_FUNCIONALIDAD)

    def test_jugador_1_gana_partida(self, monkeypatch):
        sol_file_path = 'test/test_jugador_1_gana_partida.out'
        test_path = 'test/test.out'
        with open(test_path, 'w') as file:
            sys.stdout = file
            Juego.mensaje_bienvenida()
            juego = Juego("Jugador1", "jugador2")
            # Entrada ficticia de los jugadores
            answers = iter([3, 3, 2, 2, 3, 3, 2, 2, 3, 3, 2, 2])

            # using lambda statement for mocking
            monkeypatch.setattr('builtins.input', lambda name: next(answers))
            with pytest.raises(SystemExit) as pytest_wrapped_e:
                juego.loop()
            assert pytest_wrapped_e.type == SystemExit

        check_same_output(sol_file_path, test_path,
                          CREAR_OUTPUTS_FUNCIONALIDAD)

    def test_jugador_2_gana_partida(self, monkeypatch):

        sol_file_path = 'test/test_jugador_2_gana_partida.out'
        test_path = 'test/test.out'
        with open(test_path, 'w') as file:
            sys.stdout = file
            Juego.mensaje_bienvenida()
            juego = Juego("Jugador1", "jugador2")
            # Entrada ficticia de los jugadores
            answers = iter([2, 4, 4, 2, 2, 4, 4, 2, 2, 4, 4, 2, 2, 4, 4, 2])

            # using lambda statement for mocking
            monkeypatch.setattr('builtins.input', lambda name: next(answers))

            with pytest.raises(SystemExit) as pytest_wrapped_e:
                juego.loop()
            assert pytest_wrapped_e.type == SystemExit

        check_same_output(sol_file_path, test_path,
                          CREAR_OUTPUTS_FUNCIONALIDAD)

    def test_final(self, monkeypatch):

        sol_file_path = 'test/test_final.out'
        test_path = 'test/test.out'
        with open(test_path, 'w') as file:
            sys.stdout = file
            Juego.mensaje_bienvenida()
            juego = Juego("Jugador1", "jugador2")
            # Entrada ficticia de los jugadores
            answers = iter([3, 3, 2, 5, 2, 4, 2, 5, 2, 5, 3, 4, 2, 3, 3, 2, 1])

            # using lambda statement for mocking
            monkeypatch.setattr('builtins.input', lambda name: next(answers))
            with pytest.raises(SystemExit) as e:
                juego.loop()
            assert e.type == SystemExit

        check_same_output(sol_file_path, test_path,
                          CREAR_OUTPUTS_FUNCIONALIDAD)


class Test_usuario():

    def test_a_implementar_unitario(self):
        # Eliminar la siguiente linea
        assert(False)
        
    def test_a_implementar_integracion(self):
        # Eliminar la siguiente linea
        assert(False)

    def test_a_implementar_funcional(self):
        # Eliminar la siguiente linea
        assert(False)
