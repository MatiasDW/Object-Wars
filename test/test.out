
Bienvenido a ObjectWars
Este juego es un juego para dos jugadores. Lo que un jugador realiza durante su turno es secreto, por lo que el otro jugador no debe mirar las acciones que realize el otro jugador durante su turno.

El objectivo del juego es dejar el enemigo sin puntos de vida, al empezar cada jugador dispone de 20.

Durante un turno el jugador puede comprar unidades.

Al empejar un turno cada jugador recibe 10 mondedas.

Despues de que ambos jugadores acaben sus turnos sus unidades entraran en combate. Si un jugador no tiene unidades, o son derrotadas, recibirá el daño en sus puntos de vida.

Existen tres tipologias de unidades: soldados, arqueros y caballeros. Las unidades tienen un bonus de 1.5 de daño siguiendo la siguiente jerarquía:
soldado -> arquero -> caballero -> soldado


    
    
Ronda 1
Turno de Jugador1
Tienes 0 monedas disponibles.
